python/grass/tools/support.py

```diff
@@ -0,0 +1,252 @@
+##############################################################################
+# AUTHOR(S): Vaclav Petras <wenzeslaus gmail com>
+#
+# PURPOSE:   API to call GRASS tools (modules) as Python functions
+#
+# COPYRIGHT: (C) 2025 Vaclav Petras and the GRASS Development Team
+#
+#            This program is free software under the GNU General Public
+#            License (>=v2). Read the file COPYING that comes with GRASS
+#            for details.
+##############################################################################
+
+"""
+The module provides support functionality for calling GRASS tools (modules)
+as Python functions.
+
+Do not use this module directly unless you are developing GRASS or its wrappers.
+In that case, be prepared to update your code if this module changes. This module may
+change between releases.
+"""
+
+from __future__ import annotations
+
+import json
+import shutil
+from io import StringIO
+
+
+import grass.script as gs
+
+
+class ParameterConverter:
+    def __init__(self):
+        self._numpy_inputs = {}
+        self._numpy_outputs = {}
+        self._numpy_inputs_ordered = []
+        self.stdin = None
+
+    def process_parameters(self, kwargs):
+        for key, value in kwargs.items():
+            if isinstance(value, StringIO):
+                kwargs[key] = "-"
+                self.stdin = value.getvalue()
+
+
+class ToolFunctionResolver:
+    def __init__(self, *, run_function, env):
+        self._run_function = run_function
+        self._env = env
+        self._names = None
+
+    # def __getattr__(self, name):
+    #    self.get_function(name, exception_type=AttributeError)
+
+    def get_function(self, name, exception_type):
```

Comment:
Pylint complains that `exception_type` is unused. There's two references of this function. Where it is defined, here, and in session_tools.py, in the `Tools.__getattr__`, that returns `return self._name_resolver.get_function(name, exception_type=AttributeError)`. The exception argument isn't used as far as I understand

Reply:
It was indeed unused. I added tests for that and some additional tests previously covered only indirectly through usage in Tools.